casper/src/main/scala/coop/rchain/casper/protocol/package.scala

```diff
@@ -1,63 +1,29 @@
 package coop.rchain.casper
 
 import coop.rchain.comm.protocol.routing.Packet
-import coop.rchain.comm.transport
-import scala.util.Try
 
-package object protocol {
+package object protocol extends CasperMessageProtocol {
+  import PacketTypeTag._
 
   def toCasperMessageProto(packet: Packet): Option[CasperMessageProto] =
-    packetToBlockRequest(packet) orElse
-      packetToHasBlockRequest(packet) orElse
-      packetToHasBlock(packet) orElse
-      packetToForkChoiceTipRequest(packet) orElse
-      packetToApprovedBlock(packet) orElse
-      packetToApprovedBlockRequest(packet) orElse
-      packetToBlockMessage(packet) orElse
-      packetToBlockApproval(packet) orElse
-      packetToUnapprovedBlock(packet) orElse
-      packetToNoApprovedBlockAvailable(packet)
+    PacketTypeTag
+      .withNameOption(packet.typeId)
+      .flatMap {
+        case BlockMessage             => convert[BlockMessage.type](packet)
+        case ApprovedBlock            => convert[ApprovedBlock.type](packet)
+        case ApprovedBlockRequest     => convert[ApprovedBlockRequest.type](packet)
+        case BlockRequest             => convert[BlockRequest.type](packet)
+        case HasBlockRequest          => convert[HasBlockRequest.type](packet)
+        case HasBlock                 => convert[HasBlock.type](packet)
+        case ForkChoiceTipRequest     => convert[ForkChoiceTipRequest.type](packet)
+        case BlockApproval            => convert[BlockApproval.type](packet)
+        case UnapprovedBlock          => convert[UnapprovedBlock.type](packet)
+        case NoApprovedBlockAvailable => convert[NoApprovedBlockAvailable.type](packet)
+      }
 
-  private def packetToBlockMessage(msg: Packet): Option[BlockMessageProto] =
-    convert[BlockMessageProto](msg, transport.BlockMessage, BlockMessageProto.parseFrom)
-  private def packetToApprovedBlock(msg: Packet): Option[ApprovedBlockProto] =
-    convert[ApprovedBlockProto](msg, transport.ApprovedBlock, ApprovedBlockProto.parseFrom)
-  private def packetToApprovedBlockRequest(msg: Packet): Option[ApprovedBlockRequestProto] =
-    convert[ApprovedBlockRequestProto](
-      msg,
-      transport.ApprovedBlockRequest,
-      ApprovedBlockRequestProto.parseFrom
-    )
-  def packetToBlockRequest(msg: Packet): Option[BlockRequestProto] =
-    convert[BlockRequestProto](msg, transport.BlockRequest, BlockRequestProto.parseFrom)
-  def packetToHasBlockRequest(msg: Packet): Option[HasBlockRequestProto] =
-    convert[HasBlockRequestProto](msg, transport.HasBlockRequest, HasBlockRequestProto.parseFrom)
-  def packetToHasBlock(msg: Packet): Option[HasBlockProto] =
-    convert[HasBlockProto](msg, transport.HasBlock, HasBlockProto.parseFrom)
+  @inline def convert[Tag <: PacketTypeTag](
```

Comment:
Is the explicit `@inline` needed here? Isn't this going to be inlined automatically by the compiler?

Reply:
Log story short: no. We should configure inliner and optimizer as described here: https://www.lightbend.com/blog/scala-inliner-optimizer - only then it will take effect. I'm not sure if we do that in prod, we definitely should (I benchmarked this stuff and monomorphisation is a huge win, especially with tagless-final and/or transformers)